tawazi/_dag/helpers.py

```diff
@@ -160,25 +143,103 @@ def wait_for_finished_nodes(
     return done, running, runnable_xns_ids
 
 
+async def wait_for_finished_nodes_async(
+    return_when: str,
+    graph: DiGraphEx,
+    futures: BiDict[Identifier, "asyncio.Future[Any]"],
+    done: Set["asyncio.Future[Any]"],
+    running: Set["asyncio.Future[Any]"],
+    runnable_xns_ids: Set[Identifier],
+) -> Tuple[Set["asyncio.Future[Any]"], Set["asyncio.Future[Any]"], Set[Identifier]]:
+    """Wait for the finished futures before pruning them from the graph.
+
+    Args:
+        return_when: condition for thread return
+        graph: the directed graph
+        futures: the futures to id map and reverse map
+        done: the set of finished futures
+        running: the running async-threads
+        runnable_xns_ids: the exec nodes that are available to be run
+
+    Returns:
+        finisehd futures, running futures and runnable nodes
+    """
+    if len(running) == 0:
+        return done, running, runnable_xns_ids
+    done_, running = await asyncio.wait(running, return_when=return_when)
+    done = done.union(done_)
+
+    # 1. among the finished futures:
+    #   1. checks for exceptions
+    #   2. and remove them from the graph
+    for done_future in done_:
+        future_id = futures.inverse[done_future]
+        _ = futures[future_id].result()  # raise exception by calling the future
+        logger.debug("Remove ExecNode {} from the graph", future_id)
+        runnable_xns_ids |= graph.remove_root_node(future_id)
+
+    return done, running, runnable_xns_ids
+
+
+async def to_thread_in_executor(
+    func: Callable[..., Any],
+    executor: ThreadPoolExecutor,
+    # /, # support for python3.7
+    *args: Tuple[Any],
+    **kwargs: Dict[str, Any],
+) -> "asyncio.Future[Any]":
+    """A modified copy of asyncio.to_thread.
+
+    Asynchronously run function *func* in a separate thread.
+
+    Any *args and **kwargs supplied for this function are directly passed
+    to *func*. Also, the current :class:`contextvars.Context` is propagated,
+    allowing context variables from the main thread to be accessed in the
+    separate thread.
+
+    Return a coroutine that can be awaited to get the eventual result of *func*.
+    """
+    loop = asyncio.get_running_loop()
+    ctx = contextvars.copy_context()
+    func_call = functools.partial(ctx.run, func, *args, **kwargs)
+    return await loop.run_in_executor(executor, func_call)
+
+
 ################
 # The scheduler!
 ################
-def execute(
+def sync_execute(
+    exec_nodes: StrictDict[Identifier, ExecNode],
+    results: StrictDict[Identifier, Any],
+    max_concurrency: int,
+    graph: DiGraphEx,
+) -> Tuple[
+    StrictDict[Identifier, ExecNode], StrictDict[Identifier, Any], StrictDict[Identifier, Profile]
+]:
+    """Look at the execute function for more information."""
+    return asyncio.run(
```

Comment:
why is sync execute using async execute ??

Reply:
because it is the simplest solution! I know that this is ugly but it allows us to avoid duplicating the core of tawazi (the scheduler) which is the most critical part!
I don't like it too but I couldn't find a good solution yet

Reply:
the problem is that we need to have an await inside the scheduler, otherwise we can't have async working correctly.

Reply:
it's very very weird imo, besides, I'm not sure that there are no side effects when doing that tbh

Reply:
I agree, we will see what to do later on